prow/pod-utils/gcs/upload.go

```diff
@@ -35,29 +35,48 @@ type UploadFunc func(obj *storage.ObjectHandle) error
 // Upload uploads all of the data in the
 // uploadTargets map to GCS in parallel. The map is
 // keyed on GCS path under the bucket
+// 3 attempts will be made to have an error free upload
 func Upload(bucket *storage.BucketHandle, uploadTargets map[string]UploadFunc) error {
-	errCh := make(chan error, len(uploadTargets))
-	group := &sync.WaitGroup{}
-	group.Add(len(uploadTargets))
-	for dest, upload := range uploadTargets {
-		obj := bucket.Object(dest)
-		logrus.WithField("dest", dest).Info("Queued for upload")
-		go func(f UploadFunc, obj *storage.ObjectHandle, name string) {
-			defer group.Done()
-			if err := f(obj); err != nil {
-				errCh <- err
+
+	retries := 3
+
+	for i := 0; i < retries; i++ {
+
+		errCh := make(chan error, len(uploadTargets))
+		group := &sync.WaitGroup{}
+		group.Add(len(uploadTargets))
+
+		for dest, upload := range uploadTargets {
+
+			obj := bucket.Object(dest)
+			logrus.WithField("dest", dest).Info("Queued for upload")
+			go func(f UploadFunc, obj *storage.ObjectHandle, name string) {
```

Comment:
Could we do retries local to a specific worker? Can we retry from within the goroutine?

Reply:
just updated to try this. tests pass at least. Look ok? feel free to throw any advice my way. this is all new to me :)